pkg/controller/tailoredprofile/tailoredprofile_controller.go

```diff
@@ -536,6 +657,106 @@ func (r *ReconcileTailoredProfile) getRulesFromSelections(tp *cmpv1alpha1.Tailor
+		ruleKey := types.NamespacedName{Name: selection.Name, Namespace: tp.Namespace}
+		err := r.Client.Get(context.TODO(), ruleKey, rule)
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				return nil, common.NewNonRetriableCtrlError("Fetching rule: %w", err)
+			}
+			return nil, err
+		}
+
+		// Make sure all CustomRule has ScannerType as CEL as we only support CEL at this time
+		if rule.Spec.ScannerType != cmpv1alpha1.ScannerTypeCEL {
+			return nil, common.NewNonRetriableCtrlError("CustomRule '%s' has unsupported ScannerType: %s", rule.Name, rule.Spec.ScannerType)
+		}
+
+		// Check if the CustomRule is ready (validated)
+		if rule.Status.Phase != cmpv1alpha1.CustomRulePhaseReady {
+			if rule.Status.Phase == cmpv1alpha1.CustomRulePhaseError {
+				return nil, common.NewNonRetriableCtrlError("CustomRule '%s' has validation errors: %s", rule.Name, rule.Status.ErrorMessage)
+			}
+			// If the rule is not ready yet (perhaps still being validated), return a retriable error
+			// This will cause the TailoredProfile to be requeued
+			// Note: plain errors are retriable by default in this codebase
+			return nil, fmt.Errorf("CustomRule '%s' is not ready yet", rule.Name)
+		}
+
+		rules[selection.Name] = rule
+	}
+	return rules, nil
+}
+
+// validateNoMixedRuleTypes ensures that a TailoredProfile doesn't mix CustomRules and regular Rules
+func (r *ReconcileTailoredProfile) validateNoMixedRuleTypes(tp *cmpv1alpha1.TailoredProfile, customRules map[string]*cmpv1alpha1.CustomRule) error {
+	// If we have CustomRules, check if there are any regular Rules
+	if len(customRules) > 0 {
+		hasRegularRules := false
+		var regularRuleNames []string
+		var customRuleNames []string
+
+		for _, selection := range append(tp.Spec.EnableRules, append(tp.Spec.DisableRules, tp.Spec.ManualRules...)...) {
+			// If Kind is empty or explicitly set to "Rule", it's a regular rule
+			if selection.Kind == "" || selection.Kind == cmpv1alpha1.RuleKind {
+				hasRegularRules = true
+				regularRuleNames = append(regularRuleNames, selection.Name)
+			} else if selection.Kind == cmpv1alpha1.CustomRuleKind {
+				customRuleNames = append(customRuleNames, selection.Name)
+			}
+		}
+
+		if hasRegularRules {
+			return common.NewNonRetriableCtrlError(
+				"TailoredProfile cannot mix CustomRules and regular Rules. Found %d CustomRules (%v) and %d regular Rules (%v). Please use either CustomRules or regular Rules, not both",
+				len(customRuleNames), customRuleNames, len(regularRuleNames), regularRuleNames,
+			)
+		}
+	}
+
+	// Also check the reverse case: if we have regular rules with Kind explicitly set to CustomRule
+	// This is already handled by getCustomRulesFromSelections() but let's be explicit
+
+	return nil
+}
+
+// getCustomVariablesFromSelections is for getting all avaiable variables for CustomRule
+func (r *ReconcileTailoredProfile) getCustomVariablesFromSelections(tp *cmpv1alpha1.TailoredProfile) ([]*cmpv1alpha1.Variable, error) {
```

Comment:
I'm not sure how this method would be different from getting selection for rules.

Do we have something that sets variables for the CustomRules and CEL scanner? I'm not sure I understand how those two things work well together given the flexibility we have to CEL expressions and inputs already.

Reply:
I think this is backward compatible with setValue we have, and in the future we could potentially have CustomeVariabe?